src/service/xhr-impl.js

```diff
@@ -112,19 +112,27 @@ export class Xhr {
    * @private
    */
   fetchAmpCors_(input, opt_init) {
+    const init = opt_init || {};
     input = this.getCorsUrl(this.win, input);
-    return this.fetch_(input, opt_init).then(response => {
+    // For some same origin requests, add AMP-Same-Origin: true header to allow
+    // publishers to validate that this request came from their own origin.
+    const sourceOrigin = getSourceOrigin(this.win.location.href);
```

Comment:
This is not correct. We should look not for a source origin, but the actual origin. I assume it'd be `parseUrl(...).origin`.